Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Abdellatif El Khlifi <abdellatif.elkhlifi@arm.com>

From dc2c23ba5acf17644af5ce5efc90667ce5da56c6 Mon Sep 17 00:00:00 2001
From: Morten Borup Petersen <Morten.Petersen@arm.com>
Date: Fri, 1 Mar 2019 13:29:18 +0000
Subject: [PATCH 09/20] misc: arm: add corstone700 external system harness
 driver

This commit adds a driver and device node for controlling the external
system harness of arm corstone700.
Currently, the only functionality is ioctl command: 0x0, which will
deassert the CPU_WAIT pin of the external system harness attached to the
driver.

Signed-off-by: Morten Borup Petersen <Morten.Petersen@arm.com>
Signed-off-by: Rui Miguel Silva <rui.silva@linaro.org>
---
 arch/arm/configs/corstone700_defconfig |   1 +
 drivers/misc/Kconfig                   |   1 +
 drivers/misc/Makefile                  |   1 +
 drivers/misc/arm/Kconfig               |   5 +
 drivers/misc/arm/Makefile              |   1 +
 drivers/misc/arm/extsys_ctrl.c         | 147 +++++++++++++++++++++++++
 6 files changed, 156 insertions(+)
 create mode 100644 drivers/misc/arm/Kconfig
 create mode 100644 drivers/misc/arm/Makefile
 create mode 100644 drivers/misc/arm/extsys_ctrl.c

diff --git a/arch/arm/configs/corstone700_defconfig b/arch/arm/configs/corstone700_defconfig
index 5d5244ce8278..629b2163a534 100644
--- a/arch/arm/configs/corstone700_defconfig
+++ b/arch/arm/configs/corstone700_defconfig
@@ -173,3 +173,4 @@ CONFIG_ARM_MHUv2=y
 CONFIG_RPMSG=y
 CONFIG_RPMSG_CHAR=y
 CONFIG_RPMSG_ARM=y
+CONFIG_EXTSYS_CTRL=y
\ No newline at end of file
diff --git a/drivers/misc/Kconfig b/drivers/misc/Kconfig
index 7f0d48f406e3..8d904cd46e4b 100644
--- a/drivers/misc/Kconfig
+++ b/drivers/misc/Kconfig
@@ -480,4 +480,5 @@ source "drivers/misc/cxl/Kconfig"
 source "drivers/misc/ocxl/Kconfig"
 source "drivers/misc/cardreader/Kconfig"
 source "drivers/misc/habanalabs/Kconfig"
+source "drivers/misc/arm/Kconfig"
 endmenu
diff --git a/drivers/misc/Makefile b/drivers/misc/Makefile
index c1860d35dc7e..9eec96cfb750 100644
--- a/drivers/misc/Makefile
+++ b/drivers/misc/Makefile
@@ -57,3 +57,4 @@ obj-y				+= cardreader/
 obj-$(CONFIG_PVPANIC)   	+= pvpanic.o
 obj-$(CONFIG_HABANA_AI)		+= habanalabs/
 obj-$(CONFIG_XILINX_SDFEC)	+= xilinx_sdfec.o
+obj-y				+= arm/
diff --git a/drivers/misc/arm/Kconfig b/drivers/misc/arm/Kconfig
new file mode 100644
index 000000000000..a808b5b17b8d
--- /dev/null
+++ b/drivers/misc/arm/Kconfig
@@ -0,0 +1,5 @@
+config EXTSYS_CTRL
+	tristate "Arm External System control driver"
+	help
+	  Say y here to enable support for external system control
+	  driver for the Arm Corstone-700 platform
diff --git a/drivers/misc/arm/Makefile b/drivers/misc/arm/Makefile
new file mode 100644
index 000000000000..1ca3084cf8a0
--- /dev/null
+++ b/drivers/misc/arm/Makefile
@@ -0,0 +1 @@
+obj-$(CONFIG_EXTSYS_CTRL)	+= extsys_ctrl.o
diff --git a/drivers/misc/arm/extsys_ctrl.c b/drivers/misc/arm/extsys_ctrl.c
new file mode 100644
index 000000000000..1ff384ef9c0d
--- /dev/null
+++ b/drivers/misc/arm/extsys_ctrl.c
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Arm Corstone700 external system reset control driver
+ *
+ * Copyright (C) 2019 Arm Ltd.
+ *
+ */
+
+#include <linux/fs.h>
+#include <linux/clk.h>
+#include <linux/err.h>
+#include <linux/interrupt.h>
+#include <linux/io.h>
+#include <linux/kernel.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/miscdevice.h>
+#include <linux/init.h>
+
+#define EXTSYS_DRV_NAME		"extsys_ctrl"
+#define EXTSYS_MAX_DEVS		 4
+
+#define EXTSYS_RST_SIZE		U(0x8)
+#define EXTSYS_RST_CTRL_OFF	U(0x0)
+#define EXTSYS_RST_ST_OFF	U(0x4)
+
+/* External system reset control indexes */
+#define EXTSYS_CPU_WAIT		(0x0)
+#define EXTSYS_RST_REQ		(0x1)
+
+/* External system reset status masks */
+#define EXTSYS_RST_ST_ACK_OFF	U(0x1)
+
+/* No Reset Requested */
+#define EXTSYS_RST_ST_ACK_NRR	(0x0 << EXTSYS_RST_ST_ACK_OFF)
+
+/* Reset Request Complete */
+#define EXTSYS_RST_ST_ACK_RRC	(0x2 << EXTSYS_RST_ST_ACK_OFF)
+
+/* Reset Request Unable to Complete */
+#define EXTSYS_RST_ST_ACK_RRUC	(0x3 << EXTSYS_RST_ST_ACK_OFF)
+
+/* IOCTL commands */
+#define EXTSYS_CPU_WAIT_DISABLE	0x0
+
+struct extsys_ctrl {
+	struct miscdevice miscdev;
+	void __iomem *reset_reg;
+	void __iomem *set_reg;
+};
+
+#define CLEAR_BIT(addr, index) writel(readl(addr) & ~(1UL << index), addr)
+
+static long extsys_ctrl_ioctl(struct file *f, unsigned int cmd,
+			      unsigned long arg)
+{
+	struct extsys_ctrl *extsys;
+
+	extsys = container_of(f->private_data, struct extsys_ctrl, miscdev);
+
+	switch (cmd) {
+	case EXTSYS_CPU_WAIT_DISABLE:
+		CLEAR_BIT(extsys->reset_reg, EXTSYS_CPU_WAIT);
+		break;
+
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static const struct file_operations extsys_ctrl_fops = {
+	.owner = THIS_MODULE,
+	.unlocked_ioctl = extsys_ctrl_ioctl,
+};
+
+static int extsys_ctrl_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct extsys_ctrl *extsys;
+	struct resource *res;
+	void __iomem *reset_reg;
+	void __iomem *set_reg;
+	int ret;
+
+	res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "rstreg");
+	reset_reg = devm_ioremap_resource(dev, res);
+	if (IS_ERR(reset_reg))
+		return PTR_ERR(reset_reg);
+
+	res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "streg");
+	set_reg = devm_ioremap_resource(dev, res);
+	if (IS_ERR(set_reg))
+		return PTR_ERR(set_reg);
+
+	extsys = devm_kzalloc(dev, sizeof(*extsys), GFP_KERNEL);
+	if (!extsys)
+		return -ENOMEM;
+
+	extsys->reset_reg = reset_reg;
+	extsys->set_reg = set_reg;
+
+	extsys->miscdev.minor = MISC_DYNAMIC_MINOR;
+	extsys->miscdev.name = EXTSYS_DRV_NAME;
+	extsys->miscdev.fops = &extsys_ctrl_fops;
+	extsys->miscdev.parent = dev;
+
+	ret = misc_register(&extsys->miscdev);
+	if (ret)
+		return ret;
+
+	dev_info(dev, "external system controller ready\n");
+
+	return 0;
+}
+
+static int extsys_ctrl_remove(struct platform_device *pdev)
+{
+	struct extsys_ctrl *extsys = dev_get_drvdata(&pdev->dev);
+
+	misc_deregister(&extsys->miscdev);
+
+	return 0;
+}
+
+static const struct of_device_id extsys_ctrl_match[] = {
+	{ .compatible = "arm,extsys_ctrl" },
+	{ },
+};
+MODULE_DEVICE_TABLE(of, extsys_ctrl_match);
+
+static struct platform_driver extsys_ctrl_driver = {
+	.driver = {
+		.name = EXTSYS_DRV_NAME,
+		.of_match_table = extsys_ctrl_match,
+	},
+	.probe = extsys_ctrl_probe,
+	.remove = extsys_ctrl_remove,
+};
+module_platform_driver(extsys_ctrl_driver);
+
+MODULE_LICENSE("GPL v2");
+MODULE_DESCRIPTION("Arm External System Control Driver");
+MODULE_AUTHOR("Morten Borup Petersen <morten.petersen@arm.com>");
+MODULE_AUTHOR("Rui Miguel Silva <rui.silva@linaro.org>");
-- 
2.17.1

